React/React hooks.md

```diff
@@ -732,6 +732,291 @@ Useful links:
+  }, []);
+
+  useEffect(() => {
+    if (!loading && !error && data) {
+      onSuccess();
+    }
+  }, [loading, error, data, onSuccess]);
+
+  return <div>Data: {data}</div>;
+}
+```
+
+The Problem ⚡  
+
+There are two useEffect hooks, the first one is handling the api call on the initial render and the second one will call the onSuccess function, by assuming when there is no loading, no error, but data in the state, it must have been a successful call. Makes sense right?
+
+Sure for the first call this is true and probably will never fail. But you also loose the direct connection between the action and the function that needs to be called. There is no guarantee that this case will only happen if the fetch action has succeeded and this is something we as developers really don't like.
+
+The Solution ✅  
+
+A straight forward solution would be to set the "onSuccess" function to the actual place where the call was successful:  
+
+```jsx
+function DataList({ onSuccess }) {
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState(null);
+  const [data, setData] = useState(null);
+
+  const fetchData = () => {
+    setLoading(true);
+    callApi()
+      .then((fetchedData) => {
+        setData(fetchedData);
+        onSuccess();
+      })
+      .catch((err) => setError(err))
+      .finally(() => setLoading(false));
+  };
+
+  useEffect(() => {
+    fetchData();
+  }, []);
+
+  return <div>{data}</div>;
+}
+```
+
+Now it is quite clear at first sight when the onSuccess is called, exactly in the success case of the api call.
+
+#### Single responsibility useEffects
+
+Remember the times when we only had the componentWillReceiveProps or componentDidUpdate methods to hook into the rendering process of a react component? It is bringing back dark memories and also realizing the beauty of using the useEffect hook and especially that you can have as much as you want of them.
+
+But sometimes forgetting and using a "useEffect" for several things brings back those dark memories. For example, imagine you have a component that fetches some data from the backend in some way, and also displays breadcrumbs depending on the current location.
+
+This is dangerous ❌  
+
+```jsx
+function Example(props) {
+  const location = useLocation();
+
+  const fetchData = () => {
+    /*  Calling the api */
+  };
+
+  const updateBreadcrumbs = () => {
+    /* Updating the breadcrumbs*/
+  };
+
+  useEffect(() => {
+    fetchData();
+    updateBreadcrumbs();
+  }, [location.pathname]);
+
+  return (
+    <div>
+      <BreadCrumbs />
+    </div>
+  );
+}
+```
+
+The Problem ⚡  
+
+There are two use cases, the "data-fetching" and "displaying breadcrumbs". Both are updated with an useEffect hook. This single useEffect hooks will run when the fetchData and updateBreadcrumbs functions or the location changes. The main problem is now, we also call the fetchData function when the location changes. This might be a side effect we haven't thought of.
```

Comment:
```suggestion
There are two use cases, the "data-fetching" and "displaying breadcrumbs". Both are updated with an useEffect hook. This single useEffect hooks will run when the fetchData and updateBreadcrumbs functions on the location changes. The main problem is now, we also call the fetchData function when the location changes. This might be a side effect we haven't thought of.
```

Reply:
tbh, I don't exactly understand the point you're trying to make here. It should probably be more clear that fetchData doesn't depend on the path.

Reply:
I'm still working on this one. Should have added WIP or draft.
Chapters:
#### Handling actions via useEffect 
and
#### Single responsibility useEffects
are copied from some article, I'll do my own version of examples and try to better explain the problem.
This was just the layout of the idea.